src/main/java/com/example/petbuddybackend/utils/provider/geolocation/GeolocationProvider.java

```diff
@@ -0,0 +1,82 @@
+package com.example.petbuddybackend.utils.provider.geolocation;
+
+import com.example.petbuddybackend.utils.exception.throweable.general.NotFoundException;
+import com.example.petbuddybackend.utils.provider.geolocation.dto.Coordinates;
+import com.opencagedata.jopencage.JOpenCageGeocoder;
+import com.opencagedata.jopencage.model.JOpenCageForwardRequest;
+import com.opencagedata.jopencage.model.JOpenCageResponse;
+import com.opencagedata.jopencage.model.JOpenCageResult;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Component;
+
+import java.math.BigDecimal;
+import java.math.RoundingMode;
+
+@Component
+@RequiredArgsConstructor
+public class GeolocationProvider {
+
+    private static final String ADDRESS_FORMATT = "%s, %s, %s";
+    private static final String GEOLOCATION = "Geolocation";
+
+    private final JOpenCageGeocoder jOpenCageGeocoder;
+
+    private static Long lastRequestTime = 1_577_833_200_000L;
+
+    public Coordinates getCoordinatesOfAddress(String country, String city) {
+        return getCoordinatesOfAddress(country, city, null);
+    }
+
+    public Coordinates getCoordinatesOfAddress(String country, String city, String streetName) {
+        String formattedAddress = formatAddress(country, city, streetName);
+        JOpenCageForwardRequest request = buildRequest(formattedAddress);
+        JOpenCageResponse response = sendRequest(request);
+        assertResultExists(response, formattedAddress);
+        return unpackResultToCoordinates(response);
+    }
```

Comment:
Pytanie czy rzucać wyjątek jak nic nie znajdzie. Widzę że tego w get caretakers używasz w filtrze a w teorii no i imo jak filtr jest zbyt szczegółowy to powinien pusty page zwrócić zamiast wyjątku

Reply:
Znaczy jeśli nie znajdzie wspolrzedncyh to znaczy że coś nie tak opiekun podał w formularzu chyba więc wyjątek nienajgorszy. A w przypadku geta to jest tylko do pobrania współrzędnych miasta więc nwm może tutaj rzeczywiście jak nie znajdzie to nulle wrzucić. Wtedy obsługę wyjątków do serwisu musiałbym przenieść z providera.